src/module/custom.rs

```diff
@@ -0,0 +1,279 @@
+where
+    T: CustomSection,
+{
+    #[inline]
+    fn from(a: TypedCustomSectionId<T>) -> Self {
+        a.id
+    }
+}
+
+impl Tombstone for Option<Box<dyn CustomSectionAny>> {
+    fn on_delete(&mut self) {
+        *self = None;
+    }
+}
+
+/// A collection of custom sections inside a Wasm module.
+///
+/// To add parse and emit your own custom section:
+///
+/// * Define a `MyCustomSection` type to represent your custom section.
+///
+/// * Implement the `walrus::CustomSection` trait for your `MyCustomSection`
+///   type.
+///
+/// * When working with a `walrus::Module` named `my_module`, use
+///   `my_module.customs.take_raw("my_custom_section_name")` to take ownership
+///   of the raw custom section if it is present in the Wasm module.
+///
+/// * Parse that into your own `MyCustomSection` type in whatever way is
+///   appropriate.
+///
+/// * Do whatever kinds of inspection and manipulation of your `MyCustomSection`
+///   type you need to do.
+///
+/// * Use `my_module.customs.add(my_custom_section)` to add the custom section
+///   back into the module, so `walrus` can emit the processed/updated version
+///   of the custom section.
+#[derive(Debug, Default)]
+pub struct ModuleCustomSections {
+    arena: TombstoneArena<Option<Box<dyn CustomSectionAny>>>,
+}
+
+impl ModuleCustomSections {
+    /// Add a new custom section to the module.
+    pub fn add<T>(&mut self, custom_section: T) -> TypedCustomSectionId<T>
+    where
+        T: CustomSection,
+    {
+        let id = self
+            .arena
+            .alloc(Some(Box::new(custom_section) as Box<dyn CustomSectionAny>));
+        TypedCustomSectionId {
+            id: CustomSectionId(id),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Remove a custom section from the module.
+    pub fn delete<I>(&mut self, id: I)
+    where
+        I: Into<CustomSectionId>,
+    {
+        let id = id.into().0;
+        self.arena.delete(id);
+    }
+
+    /// Take a raw, unparsed custom section out of this module.
+    pub fn take_raw(&mut self, name: &str) -> Option<RawCustomSection> {
+        let id = self
+            .arena
+            .iter()
+            .filter(|(_id, s)| {
+                if let Some(s) = s {
+                    s.as_any().is::<RawCustomSection>() && s.name() == name
+                } else {
+                    false
+                }
+            })
+            .map(|(id, _)| id)
+            .next()?;
+        let section = self.arena[id].take().unwrap();
+        self.arena.delete(id);
+        let raw = section.into_any().downcast::<RawCustomSection>().unwrap();
+        Some(*raw)
+    }
+
+    /// Get a shared reference to a custom section that is in this
+    /// `ModuleCustomSections`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the custom section associated with the given `id` has been
+    /// deleted or taken.
+    pub fn get<T>(&self, id: TypedCustomSectionId<T>) -> &T
```

Comment:
Following the precedent of std it may be best to remove `try_*` and have this family of functions return `Option`?

Reply:
And then all the things that are currently `get` should be `Index` impls?

Reply:
In theory yeah, although I'd be fine leaving the `Index` impls for later.